### ISSUE TITLE
DROID forward-dynamics SFT loss  does not converge, it seems does not learn action-video correspondence

### ISSUE THREAD

**Issue by YingQQ666** [NVIDIA/cosmos-framework#175]
I try to  SFT the forward-dynamics model in DROID , from nano ckpt

> initial three-camera observation +  actions → future frames

keys_to_skip_loading = [
    "net_ema.",
    "lora_",
    "action_pos_embed", 
    "action_modality_embed",
    "action2llm",
    "llm2action",
]
and
keys_to_select = [
    "moe_gen",
    "time_embedder",
    "vae2llm",
    "llm2vae",
    "action2llm",
    "action_modality_embed",
    "action_pos_embed"
]
The future-video flow-matching loss oscillates around 0.2 and fails to converge, action perturbations change the predicted velocity field, and gradients reach `action2llm`. However, with matched validation samples and matched diffusion noise, correct actions produce almost the same loss and generated trajectory as shuffled valid actions.

## Setup

- Base checkpoint: public `nvidia/Cosmos3-Nano`, converted to DCP
- Hardware: 8 × A100 80 GB, BF16, FSDP shard degree 8
- Dataset: Cosmos3-DROID success + failure
- Input: 33 frames at 15 FPS
- Condition: frame 0 and 32 actions
- Target: 32 future frames
- Action: raw 8D absolute joint position, padded to 64D
- State condition: disabled
- Viewpoint: three-camera `concat_view`
- Global batch per optimizer step: 256
- Prompt: `Generate the future video conditioned on the initial visual observation and the provided action sequence.`

The evaluated iteration-1000 checkpoint used:

```yaml
optimizer_lr: 2.0e-5
action_lr_multiplier: 1.0
scheduler_cycle: 10000
warmup_steps: 250
skipped_on_base_load:
  - action2llm
```

## Counterfactual validation

I evaluated the same 640 fixed validation windows using:

1. Correct actions
2. Valid actions globally shuffled across the distributed batch
3. Zero action values

All branches reused the same images, targets, sigma, noise, and prompt. The checkpoint was fully loaded during evaluation.

### Iteration 1000, noise seed 0

| Condition | Future-video FM loss | Relative to correct |
|---|---:|---:|
| Correct | `0.16763957` | — |
| Shuffled | `0.16763823` | `-0.000808%` |
| Zero | `0.16769676` | `+0.034118%` |

Across noise seeds 0, 1, and 2:

- Mean shuffled penalty: `+0.001860%`
- Mean zero penalty: `+0.034967%`
- `RMS(v_shuffled - v_correct)`: approximately `0.0176`
- `RMS(v_zero - v_correct)`: approximately `0.030–0.031`

Thus, action values affect the predicted velocity, but correct action-video pairing provides almost no advantage.

Between iterations 500 and 1000:

- Correct loss improved from `0.16834874` to `0.16763957`
- Zero-action penalty decreased from `0.0873%` to `0.0341%`

Same-seed generated videos also showed very similar correct and shuffled trajectories.

## Gradient-path check

I later changed the local action LR multiplier to 10× and ran one 8-GPU optimizer-step diagnostic loaded from iteration 1000. This was only a gradient-path test, not a full 10× training run.

Results:

- All real 8D action tensors received finite gradients
- All eight channels had nonzero gradient norms
- `action2llm` was trainable, present in the optimizer, and updated
- Gradient energy was routed to DROID domain 8
- Raw 8D columns had nonzero gradients
- All 56 padded columns had zero gradient
- The effective action LR was confirmed to be 10× the video-connector LR

This appears to rule out a detached graph, frozen parameters, wrong optimizer selection, domain-routing error, or padding leak.

## Additional observations

When fully loading the public Nano action weights and feeding the same 8D joint-position actions:

- Correct loss: `0.17870769`
- Shuffled: `+0.1174%` relative to correct
- Zero loss: `0.16864462`, or `-5.6310%`

This suggests that the public checkpoint may expect a different action representation.

I also measured a consistent 3–4 frame delay between commanded joint positions and measured joint-state response. The current sequence plan nominally aligns `action[0]` with future frame 1 and provides no pre-window action history.

## Questions

1. Is there an official DROID forward-dynamics SFT recipe?
2. What action representation does the public Nano action connector expect for DROID/Franka: 8D joint positions, 10D Cartesian actions, or another schema?
3. When changing action representation, which parameters should be reinitialized: `action2llm.fc`, its bias, and/or `action_modality_embed`?
4. Should forward dynamics include action history or compensate for control latency?
5. Is future-video flow matching alone expected to learn correct action pairing, or is an auxiliary action-relevance objective recommended?
6. Should most of the video backbone initially be frozen while learning a new action connector?
7. What global batch size and training budget would you recommend?

I can provide the local recipe, dataset adapter, and matched-noise evaluation patch if useful.

## Environment
- Ubuntu 22.04
- Python 3.13.5
- PyTorch `2.10.0+cu130`
- 8 × NVIDIA A100 80 GB
